src/sources/datadog/agent/logs.rs

```diff
@@ -0,0 +1,144 @@
+                }
+            },
+        )
+        .boxed()
+}
+
+pub(crate) fn decode_log_body(
+    body: Bytes,
+    api_key: Option<Arc<str>>,
+    source: &DatadogAgentSource,
+) -> Result<Vec<Event>, ErrorMessage> {
+    if body.is_empty() {
+        // The datadog agent may send an empty payload as a keep alive
+        debug!(
+            message = "Empty payload ignored.",
+            internal_log_rate_secs = 30
+        );
+        return Ok(Vec::new());
+    }
+
+    let messages: Vec<LogMsg> = serde_json::from_slice(&body).map_err(|error| {
+        ErrorMessage::new(
+            StatusCode::BAD_REQUEST,
+            format!("Error parsing JSON: {:?}", error),
+        )
+    })?;
+
+    let now = Utc::now();
+    let mut decoded = Vec::new();
+
+    for message in messages {
+        let mut decoder = source.decoder.clone();
+        let mut buffer = BytesMut::new();
+        buffer.put(message.message);
+        loop {
+            match decoder.decode_eof(&mut buffer) {
+                Ok(Some((events, _byte_size))) => {
+                    for mut event in events {
+                        if let Event::Log(ref mut log) = event {
+                            log.try_insert_flat("status", message.status.clone());
+                            log.try_insert_flat("timestamp", message.timestamp);
+                            log.try_insert_flat("hostname", message.hostname.clone());
+                            log.try_insert_flat("service", message.service.clone());
+                            log.try_insert_flat("ddsource", message.ddsource.clone());
+                            log.try_insert_flat("ddtags", message.ddtags.clone());
+                            log.try_insert_flat(
+                                source.log_schema_source_type_key,
+                                Bytes::from("datadog_agent"),
+                            );
+                            log.try_insert_flat(source.log_schema_timestamp_key, now);
+                            if let Some(k) = &api_key {
+                                log.metadata_mut().set_datadog_api_key(Some(Arc::clone(k)));
+                            }
+                        }
+
+                        decoded.push(event);
+                    }
+                }
+                Ok(None) => break,
+                Err(error) => {
+                    // Error is logged by `crate::codecs::Decoder`, no further
+                    // handling is needed here.
+                    if !error.can_continue() {
+                        break;
+                    }
+                }
+            }
+        }
+    }
+
+    emit!(&EventsReceived {
+        byte_size: decoded.size_of(),
+        count: decoded.len(),
+    });
+
+    Ok(decoded)
+}
+
+// https://github.com/DataDog/datadog-agent/blob/a33248c2bc125920a9577af1e16f12298875a4ad/pkg/logs/processor/json.go#L23-L49
+#[derive(Deserialize, Clone, Serialize, Debug)]
+#[serde(deny_unknown_fields)]
```

Comment:
I imagine this was copy/pasted, but it does seem to me like we could be more lenient here and just ignore undefined fields. That way, if the agent adds a field, it'll still work with Vector even if Vector doesn't recognize it yet.

Reply:
Yep this looks like a safe update that could limit breaking changes to break everything. Will remove.